util/lua/lua.go

```diff
@@ -253,16 +253,18 @@ func cleanReturnedObj(newObj, obj map[string]interface{}) map[string]interface{}
 func cleanReturnedArray(newObj, obj []interface{}) []interface{} {
 	arrayToReturn := newObj
 	for i := range newObj {
-		switch newValue := newObj[i].(type) {
-		case map[string]interface{}:
-			if oldValue, ok := obj[i].(map[string]interface{}); ok {
-				convertedMap := cleanReturnedObj(newValue, oldValue)
-				arrayToReturn[i] = convertedMap
-			}
-		case []interface{}:
-			if oldValue, ok := obj[i].([]interface{}); ok {
-				convertedMap := cleanReturnedArray(newValue, oldValue)
-				arrayToReturn[i] = convertedMap
+		if len(obj) > i {
```

Comment:
If we stop handling `newObj` when we reach a point where `obj` has no more entries, then we'll fail to convert empty arrays after `i` in `newObj` to empty maps where appropriate. To reproduce the issue, make these changes to your new tests:

```diff
diff --git a/util/lua/lua_test.go b/util/lua/lua_test.go
index 97971241c..9d79eae98 100644
--- a/util/lua/lua_test.go
+++ b/util/lua/lua_test.go
@@ -780,7 +780,6 @@ spec:
 
 const addToSyncWindowAction = `
 newSyncWindow = {}
-newSyncWindow.duration = "1h"
 if obj.spec.syncWindows == nil then
   obj.spec.syncWindows = {}
 end
@@ -796,7 +795,7 @@ metadata:
 spec:
   destinations: {}
   syncWindows:
-  - duration: 1h
+  - {}
 `
 
 func TestAddToArrayNil(t *testing.T) {
```

I think the fix would be to pull the desired type from the first available item of `obj` (if there is any) and then convert _all_ `newObj` items to that type. I believe heterogeneous list item types is a safe assumption.

Reply:
Maybe something like this:

```diff
diff --git a/util/lua/lua.go b/util/lua/lua.go
index 55af8773b..ac08b852a 100644
--- a/util/lua/lua.go
+++ b/util/lua/lua.go
@@ -252,19 +252,26 @@ func cleanReturnedObj(newObj, obj map[string]interface{}) map[string]interface{}
 // empty array conversions.
 func cleanReturnedArray(newObj, obj []interface{}) []interface{} {
 	arrayToReturn := newObj
+	if len(obj) == 0 {
+		// We can't infer types from an empty array, so we return the new array as is and hope the types are correct.
+		// In effect, this means that actions can't safely add empty objects to empty arrays, because we won't know
+		// that the array is supposed to contain objects. That's an acceptable bug for now.
+		return arrayToReturn
+	}
 	for i := range newObj {
-		if len(obj) > i {
-			switch newValue := newObj[i].(type) {
-			case map[string]interface{}:
-				if oldValue, ok := obj[i].(map[string]interface{}); ok {
-					convertedMap := cleanReturnedObj(newValue, oldValue)
-					arrayToReturn[i] = convertedMap
-				}
-			case []interface{}:
-				if oldValue, ok := obj[i].([]interface{}); ok {
-					convertedMap := cleanReturnedArray(newValue, oldValue)
-					arrayToReturn[i] = convertedMap
-				}
+		switch newValue := newObj[0].(type) {
+		// We always use the first item of obj to infer the desired type of items in newObj. We do this in case the new
+		// array is longer than the old array, to avoid an index out of bounds error. We can safely assume that the
+		// first item of obj is of the same type as the rest of the items in obj.
+		case map[string]interface{}:
+			if oldValue, ok := obj[0].(map[string]interface{}); ok {
+				convertedMap := cleanReturnedObj(newValue, oldValue)
+				arrayToReturn[i] = convertedMap
+			}
+		case []interface{}:
+			if oldValue, ok := obj[0].([]interface{}); ok {
+				convertedMap := cleanReturnedArray(newValue, oldValue)
+				arrayToReturn[i] = convertedMap
 			}
 		}
 	}
```